edb/pgsql/compiler/dml.py

```diff
@@ -839,33 +838,99 @@ def compile_policy_check(
+                        args=[a, pgast.BooleanConstant(val=str(b))],
+                    ),
+                    pgast.StringConstant(val='insufficient_privilege'),
+                    pgast.NamedFuncArg(
+                        name='msg',
+                        val=msg,
+                    ),
+                    pgast.NamedFuncArg(
+                        name='table',
+                        val=pgast.StringConstant(val=str(typeref.id)),
+                    ),
+                ],
+            )
 
-        op = 'insert' if isinstance(ir_stmt, irast.InsertStmt) else 'update'
+        if isinstance(ir_stmt, irast.InsertStmt):
+            op = 'insert'
+        else:
+            op = 'update'
         msg = f'access policy violation on {op} of {typeref.name_hint}'
-        maybe_raise = pgast.FuncCall(
-            name=('edgedb', 'raise_on_null'),
-            args=[
-                pgast.FuncCall(
-                    name=('nullif',),
-                    args=[cond_ref, pgast.BooleanConstant(val='false')],
-                ),
-                pgast.StringConstant(val='insufficient_privilege'),
-                pgast.NamedFuncArg(
-                    name='msg',
-                    val=pgast.StringConstant(val=msg),
-                ),
-                pgast.NamedFuncArg(
-                    name='table',
-                    val=pgast.StringConstant(val=str(typeref.id)),
-                ),
-            ],
-        )
+
+        # allow
+        if allow:
+            allow_conds = (cond for _, cond in allow)
+            allow_expr = astutils.extend_binop(None, *allow_conds, op='OR')
+        else:
+            allow_expr = pgast.BooleanConstant(val='false')
+
+        allow_msgs = (pol.error_msg for pol, _ in allow if pol.error_msg)
+        if allow_msgs:
+            hint = 'none of these allow policies match: '
+            hint += ', '.join(allow_msgs)
+            allow_msg = f'{msg} ({hint})'
+        else:
+            allow_msg = f'{msg} (no allow policies)'
+        allow_msg_expr = pgast.StringConstant(val=allow_msg)
+
         ictx.rel.target_list.append(
-            pgast.ResTarget(name='error', val=maybe_raise)
+            pgast.ResTarget(
+                name=f'error_allow',
+                val=raise_if(allow_expr, False, msg=allow_msg_expr),
+            )
         )
 
+        # deny
+        if deny:
+            deny_conds = (cond for _, cond in deny)
+            deny_expr = astutils.extend_binop(None, *deny_conds, op='OR')
+
+            deny_messages = [
+                (
+                    pgast.StringConstant(
+                        val=f'denied by policy {pol.error_msg}'
+                    ),
+                    cond,
+                )
+                for pol, cond in deny
+                if pol.error_msg
+            ]
+            deny_message = astutils.conditional_string_agg(deny_messages)
+            if deny_message:
+                deny_message = astutils.extend_concat(
```

Comment:
Once the errmessage stuff is there, we'll need to handle the case where it errors but nothing with an error message failed